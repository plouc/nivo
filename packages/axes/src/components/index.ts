export * from './Axes'
export * from './Axis'
export * from './AxisTick'
export * from './Grid'
export * from './GridLine'
export * from './GridLines'
