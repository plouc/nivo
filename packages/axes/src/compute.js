/*
 * This file is part of the nivo project.
 *
 * Copyright 2016-present, RaphaÃ«l Benitte.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import {
    timeMillisecond,
    utcMillisecond,
    timeSecond,
    utcSecond,
    timeMinute,
    utcMinute,
    timeHour,
    utcHour,
    timeDay,
    utcDay,
    timeWeek,
    utcWeek,
    timeSunday,
    utcSunday,
    timeMonday,
    utcMonday,
    timeTuesday,
    utcTuesday,
    timeWednesday,
    utcWednesday,
    timeThursday,
    utcThursday,
    timeFriday,
    utcFriday,
    timeSaturday,
    utcSaturday,
    timeMonth,
    utcMonth,
    timeYear,
    utcYear,
} from 'd3-time'
import { timeFormat } from 'd3-time-format'
import { format as d3Format } from 'd3-format'
import { textPropsByEngine } from '@nivo/core'

export const centerScale = scale => {
    const bandwidth = scale.bandwidth()

    if (bandwidth === 0) return scale

    let offset = bandwidth / 2
    if (scale.round()) {
        offset = Math.round(offset)
    }

    return d => scale(d) + offset
}

const timeByType = {
    millisecond: [timeMillisecond, utcMillisecond],
    second: [timeSecond, utcSecond],
    minute: [timeMinute, utcMinute],
    hour: [timeHour, utcHour],
    day: [timeDay, utcDay],
    week: [timeWeek, utcWeek],
    sunday: [timeSunday, utcSunday],
    monday: [timeMonday, utcMonday],
    tuesday: [timeTuesday, utcTuesday],
    wednesday: [timeWednesday, utcWednesday],
    thursday: [timeThursday, utcThursday],
    friday: [timeFriday, utcFriday],
    saturday: [timeSaturday, utcSaturday],
    month: [timeMonth, utcMonth],
    year: [timeYear, utcYear],
}

const timeTypes = Object.keys(timeByType)
const timeIntervalRegexp = new RegExp(`^every\\s*(\\d+)?\\s*(${timeTypes.join('|')})s?$`, 'i')

const isInteger = value =>
    typeof value === 'number' && isFinite(value) && Math.floor(value) === value

export const getScaleTicks = (scale, spec) => {
    // specific values
    if (Array.isArray(spec)) {
        return spec
    }

    // continuous scales
    if (scale.ticks) {
        // default behaviour
        if (spec === undefined) {
            return scale.ticks()
        }

        // specific tick count
        if (isInteger(spec)) {
            return scale.ticks(spec)
        }

        if (typeof spec === 'string') {
            // time interval
            const matches = spec.match(timeIntervalRegexp)
            if (matches) {
                // UTC is used as it's more predictible
                // however local time could be used too
                // let's see how it fits users' requirements
                const timeType = timeByType[matches[2]][scale.useUTC ? 1 : 0]
                if (matches[1] === undefined) {
                    return scale.ticks(timeType)
                }

                return scale.ticks(timeType.every(Number(matches[1])))
            }

            throw new Error(`Invalid tickValues: ${spec}`)
        }
    }

    // non linear scale default
    return scale.domain()
}

export const computeCartesianTicks = ({
    axis,
    scale,
    ticksPosition,
    tickValues,
    tickSize,
    tickPadding,
    tickRotation,
    engine = 'svg',
}) => {
    const values = getScaleTicks(scale, tickValues)

    const textProps = textPropsByEngine[engine]

    const position = scale.bandwidth ? centerScale(scale) : scale
    const line = { lineX: 0, lineY: 0 }
    const text = { textX: 0, textY: 0 }

    let translate
    let textAlign = textProps.align.center
    let textBaseline = textProps.baseline.center

    if (axis === 'x') {
        translate = d => ({ x: position(d), y: 0 })

        line.lineY = tickSize * (ticksPosition === 'after' ? 1 : -1)
        text.textY = (tickSize + tickPadding) * (ticksPosition === 'after' ? 1 : -1)

        if (ticksPosition === 'after') {
            textBaseline = textProps.baseline.top
        } else {
            textBaseline = textProps.baseline.bottom
        }

        if (tickRotation === 0) {
            textAlign = textProps.align.center
        } else if (
            (ticksPosition === 'after' && tickRotation < 0) ||
            (ticksPosition === 'before' && tickRotation > 0)
        ) {
            textAlign = textProps.align.right
            textBaseline = textProps.baseline.center
        } else if (
            (ticksPosition === 'after' && tickRotation > 0) ||
            (ticksPosition === 'before' && tickRotation < 0)
        ) {
            textAlign = textProps.align.left
            textBaseline = textProps.baseline.center
        }
    } else {
        translate = d => ({ x: 0, y: position(d) })

        line.lineX = tickSize * (ticksPosition === 'after' ? 1 : -1)
        text.textX = (tickSize + tickPadding) * (ticksPosition === 'after' ? 1 : -1)

        if (ticksPosition === 'after') {
            textAlign = textProps.align.left
        } else {
            textAlign = textProps.align.right
        }
    }

    const ticks = values.map(value => ({
        key: value,
        value,
        ...translate(value),
        ...line,
        ...text,
    }))

    return {
        ticks,
        textAlign,
        textBaseline,
    }
}

export const getFormatter = (format, scale) => {
    if (!format || typeof format === 'function') return format

    if (scale.type === 'time') {
        const f = timeFormat(format)
        return d => f(new Date(d))
    }

    return d3Format(format)
}

export const computeGridLines = ({ width, height, scale, axis, values: _values }) => {
    const lineValues = Array.isArray(_values) ? _values : undefined
    const lineCount = isInteger(_values) ? _values : undefined

    const values = lineValues || getScaleTicks(scale, lineCount)

    const position = scale.bandwidth ? centerScale(scale) : scale

    let lines
    if (axis === 'x') {
        lines = values.map(v => ({
            key: `${v}`,
            x1: position(v),
            x2: position(v),
            y1: 0,
            y2: height,
        }))
    } else if (axis === 'y') {
        lines = values.map(v => ({
            key: `${v}`,
            x1: 0,
            x2: width,
            y1: position(v),
            y2: position(v),
        }))
    }

    return lines
}
