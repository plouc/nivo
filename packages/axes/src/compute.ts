import {
    CountableTimeInterval,
    timeMillisecond,
    utcMillisecond,
    timeSecond,
    utcSecond,
    timeMinute,
    utcMinute,
    timeHour,
    utcHour,
    timeWeek,
    utcWeek,
    timeSunday,
    utcSunday,
    timeMonday,
    utcMonday,
    timeTuesday,
    utcTuesday,
    timeWednesday,
    utcWednesday,
    timeThursday,
    utcThursday,
    timeFriday,
    utcFriday,
    timeSaturday,
    utcSaturday,
    timeMonth,
    utcMonth,
    timeYear,
    utcYear,
    timeInterval,
} from 'd3-time'
import { timeFormat } from 'd3-time-format'
import { format as d3Format } from 'd3-format'
// @ts-ignore
import { textPropsByEngine } from '@nivo/core'
import {
    AxisValue,
    Point,
    TicksSpec,
    AnyScale,
    ScaleWithBandwidth,
    ValueFormatter,
    Line,
} from './types'

export const centerScale = <Value>(scale: ScaleWithBandwidth) => {
    const bandwidth = scale.bandwidth()

    if (bandwidth === 0) return scale

    let offset = bandwidth / 2
    if (scale.round()) {
        offset = Math.round(offset)
    }

    return <T extends Value>(d: T) => (scale(d) ?? 0) + offset
}

const timeDay = timeInterval(
    date => date.setHours(0, 0, 0, 0),
    (date, step) => date.setDate(date.getDate() + step),
    (start, end) => (end.getTime() - start.getTime()) / 864e5,
    date => Math.floor(date.getTime() / 864e5)
)

const utcDay = timeInterval(
    date => date.setUTCHours(0, 0, 0, 0),
    (date, step) => date.setUTCDate(date.getUTCDate() + step),
    (start, end) => (end.getTime() - start.getTime()) / 864e5,
    date => Math.floor(date.getTime() / 864e5)
)

const timeByType: Record<string, [CountableTimeInterval, CountableTimeInterval]> = {
    millisecond: [timeMillisecond, utcMillisecond],
    second: [timeSecond, utcSecond],
    minute: [timeMinute, utcMinute],
    hour: [timeHour, utcHour],
    day: [timeDay, utcDay],
    week: [timeWeek, utcWeek],
    sunday: [timeSunday, utcSunday],
    monday: [timeMonday, utcMonday],
    tuesday: [timeTuesday, utcTuesday],
    wednesday: [timeWednesday, utcWednesday],
    thursday: [timeThursday, utcThursday],
    friday: [timeFriday, utcFriday],
    saturday: [timeSaturday, utcSaturday],
    month: [timeMonth, utcMonth],
    year: [timeYear, utcYear],
}

const timeTypes = Object.keys(timeByType)
const timeIntervalRegexp = new RegExp(`^every\\s*(\\d+)?\\s*(${timeTypes.join('|')})s?$`, 'i')

const isInteger = (value: unknown): value is number =>
    typeof value === 'number' && isFinite(value) && Math.floor(value) === value

const isArray = <T>(value: unknown): value is T[] => Array.isArray(value)

export const getScaleTicks = <Value extends AxisValue>(
    scale: AnyScale,
    spec?: TicksSpec<Value>
) => {
    // specific values
    if (Array.isArray(spec)) {
        return spec
    }

    if (typeof spec === 'string' && 'useUTC' in scale) {
        // time interval
        const matches = spec.match(timeIntervalRegexp)

        if (matches) {
            const [, amount, type] = matches
            // UTC is used as it's more predictible
            // however local time could be used too
            // let's see how it fits users' requirements
            const timeType = timeByType[type][scale.useUTC ? 1 : 0]

            if (type === 'day') {
                const [start, originalStop] = scale.domain()
                const stop = new Date(originalStop)

                // Set range to include last day in the domain since `interval.range` function is exclusive stop
                stop.setDate(stop.getDate() + 1)

                return timeType.every(Number(amount ?? 1))?.range(start, stop) ?? []
            }

            if (amount === undefined) {
                return scale.ticks(timeType)
            }

            const interval = timeType.every(Number(amount))

            if (interval) {
                return scale.ticks(interval)
            }
        }

        throw new Error(`Invalid tickValues: ${spec}`)
    }

    // continuous scales
    if ('ticks' in scale) {
        // default behaviour
        if (spec === undefined) {
            return scale.ticks()
        }

        // specific tick count
        if (isInteger(spec)) {
            return scale.ticks(spec)
        }
    }

    // non linear scale default
    return scale.domain()
}

export const computeCartesianTicks = <Value extends AxisValue>({
    axis,
    scale,
    ticksPosition,
    tickValues,
    tickSize,
    tickPadding,
    tickRotation,
    engine = 'svg',
}: {
    axis: 'x' | 'y'
    scale: AnyScale
    ticksPosition?: 'after' | 'before'
    tickValues?: TicksSpec<Value>
    tickSize: number
    tickPadding: number
    tickRotation: number
    engine?: 'svg' | 'canvas'
}) => {
    const values = getScaleTicks(scale, tickValues)

    const textProps = textPropsByEngine[engine]

    const position = 'bandwidth' in scale ? centerScale(scale) : scale
    const line = { lineX: 0, lineY: 0 }
    const text = { textX: 0, textY: 0 }

    const isRTL = typeof document === 'object' ? document.dir === 'rtl' : false
    let translate: (value: Value) => Point
    let textAlign: CanvasTextAlign = textProps.align.center
    let textBaseline: CanvasTextBaseline = textProps.baseline.center

    if (axis === 'x') {
        translate = d => ({ x: position(d) ?? 0, y: 0 })

        line.lineY = tickSize * (ticksPosition === 'after' ? 1 : -1)
        text.textY = (tickSize + tickPadding) * (ticksPosition === 'after' ? 1 : -1)

        if (ticksPosition === 'after') {
            textBaseline = textProps.baseline.top
        } else {
            textBaseline = textProps.baseline.bottom
        }

        if (tickRotation === 0) {
            textAlign = textProps.align.center
        } else if (
            (ticksPosition === 'after' && tickRotation < 0) ||
            (ticksPosition === 'before' && tickRotation > 0)
        ) {
            textAlign = textProps.align[isRTL ? 'left' : 'right']
            textBaseline = textProps.baseline.center
        } else if (
            (ticksPosition === 'after' && tickRotation > 0) ||
            (ticksPosition === 'before' && tickRotation < 0)
        ) {
            textAlign = textProps.align[isRTL ? 'right' : 'left']
            textBaseline = textProps.baseline.center
        }
    } else {
        translate = d => ({ x: 0, y: position(d) ?? 0 })

        line.lineX = tickSize * (ticksPosition === 'after' ? 1 : -1)
        text.textX = (tickSize + tickPadding) * (ticksPosition === 'after' ? 1 : -1)

        if (ticksPosition === 'after') {
            textAlign = textProps.align.left
        } else {
            textAlign = textProps.align.right
        }
    }

    const ticks = values.map(value => ({
        key: typeof value === 'number' || typeof value === 'string' ? value : `${value}`,
        value,
        ...translate(value),
        ...line,
        ...text,
    }))

    return {
        ticks,
        textAlign,
        textBaseline,
    }
}

export const getFormatter = <Value extends AxisValue>(
    format: string | ValueFormatter<Value> | undefined,
    scale: AnyScale
): ValueFormatter<Value> | undefined => {
    if (typeof format === 'undefined' || typeof format === 'function') return format

    if (scale.type === 'time') {
        const formatter = timeFormat(format)

        return (d => formatter(d instanceof Date ? d : new Date(d))) as ValueFormatter<Value>
    }

    return (d3Format(format) as unknown) as ValueFormatter<Value>
}

export const computeGridLines = <Value extends AxisValue>({
    width,
    height,
    scale,
    axis,
    values: _values,
}: {
    width: number
    height: number
    scale: AnyScale
    axis: 'x' | 'y'
    values?: TicksSpec<Value>
}) => {
    const lineValues = isArray<number>(_values) ? _values : undefined
    const values = lineValues || getScaleTicks(scale, _values)
    const position = 'bandwidth' in scale ? centerScale(scale) : scale

    const lines: Line[] =
        axis === 'x'
            ? values.map(value => ({
                  key: `${value}`,
                  x1: position(value) ?? 0,
                  x2: position(value) ?? 0,
                  y1: 0,
                  y2: height,
              }))
            : values.map(value => ({
                  key: `${value}`,
                  x1: 0,
                  x2: width,
                  y1: position(value) ?? 0,
                  y2: position(value) ?? 0,
              }))

    return lines
}
