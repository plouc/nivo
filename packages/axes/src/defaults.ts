import { AxisProps } from './types'

export const defaultAxisProps: Pick<
    Required<AxisProps>,
    'tickSize' | 'tickPadding' | 'tickRotation' | 'legendPosition' | 'legendOffset'
> = {
    tickSize: 5,
    tickPadding: 5,
    tickRotation: 0,
    legendPosition: 'middle',
    legendOffset: 0,
}
