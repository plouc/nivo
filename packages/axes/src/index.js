/*
 * This file is part of the nivo project.
 *
 * Copyright 2016-present, RaphaÃ«l Benitte.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
export { default as Axes } from './components/Axes'
export { default as Axis } from './components/Axis'
export { default as Grid } from './components/Grid'
export * from './canvas'
export * from './props'
