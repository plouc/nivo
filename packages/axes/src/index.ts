export * from './components'
export * from './canvas'
export * from './props'
export * from './types'
