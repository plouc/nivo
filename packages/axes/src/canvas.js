/*
 * This file is part of the nivo project.
 *
 * Copyright 2016-present, RaphaÃ«l Benitte.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import { degreesToRadians } from './utils'
import { computeCartesianTicks, getFormatter, computeGridLines } from './compute'

export const renderAxisToCanvas = (
    ctx,
    {
        axis,
        scale,
        x = 0,
        y = 0,
        length,

        ticksPosition,
        tickValues,
        tickSize = 5,
        tickPadding = 5,
        tickRotation = 0,
        format,

        legend,
        legendPosition = 'end',
        legendOffset = 0,

        theme,
    }
) => {
    const { ticks, textAlign, textBaseline } = computeCartesianTicks({
        axis,
        scale,
        ticksPosition,
        tickValues,
        tickSize,
        tickPadding,
        tickRotation,
        engine: 'canvas',
    })

    ctx.save()
    ctx.translate(x, y)

    ctx.textAlign = textAlign
    ctx.textBaseline = textBaseline
    ctx.font = `${theme.axis.ticks.text.fontSize}px ${theme.axis.ticks.text.fontFamily}`

    if (theme.axis.domain.line.strokeWidth > 0) {
        ctx.lineWidth = theme.axis.domain.line.strokeWidth
        ctx.lineCap = 'square'
        ctx.strokeStyle = theme.axis.domain.line.stroke
        ctx.beginPath()
        ctx.moveTo(0, 0)
        ctx.lineTo(axis === 'x' ? length : 0, axis === 'x' ? 0 : length)
        ctx.stroke()
    }

    ticks.forEach(tick => {
        if (theme.axis.ticks.line.strokeWidth > 0) {
            ctx.lineWidth = theme.axis.ticks.line.strokeWidth
            ctx.lineCap = 'square'
            ctx.strokeStyle = theme.axis.ticks.line.stroke
            ctx.beginPath()
            ctx.moveTo(tick.x, tick.y)
            ctx.lineTo(tick.x + tick.lineX, tick.y + tick.lineY)
            ctx.stroke()
        }

        const value = format !== undefined ? format(tick.value) : tick.value

        ctx.save()
        ctx.translate(tick.x + tick.textX, tick.y + tick.textY)
        ctx.rotate(degreesToRadians(tickRotation))
        ctx.fillStyle = theme.axis.ticks.text.fill
        ctx.fillText(value, 0, 0)
        ctx.restore()
    })

    if (legend !== undefined) {
        let legendX = 0
        let legendY = 0
        let legendRotation = 0
        let textAlign

        if (axis === 'y') {
            legendRotation = -90
            legendX = legendOffset
            if (legendPosition === 'start') {
                textAlign = 'start'
                legendY = length
            } else if (legendPosition === 'middle') {
                textAlign = 'center'
                legendY = length / 2
            } else if (legendPosition === 'end') {
                textAlign = 'end'
            }
        } else {
            legendY = legendOffset
            if (legendPosition === 'start') {
                textAlign = 'start'
            } else if (legendPosition === 'middle') {
                textAlign = 'center'
                legendX = length / 2
            } else if (legendPosition === 'end') {
                textAlign = 'end'
                legendX = length
            }
        }

        ctx.translate(legendX, legendY)
        ctx.rotate(degreesToRadians(legendRotation))
        ctx.font = `${
            theme.axis.legend.text.fontWeight ? `${theme.axis.legend.text.fontWeight} ` : ''
        }${theme.axis.legend.text.fontSize}px ${theme.axis.legend.text.fontFamily}`
        ctx.fillStyle = theme.axis.legend.text.fill
        ctx.textAlign = textAlign
        ctx.textBaseline = 'middle'
        ctx.fillText(legend, 0, 0)
    }

    ctx.restore()
}

const positions = ['top', 'right', 'bottom', 'left']

export const renderAxesToCanvas = (
    ctx,
    {
        xScale,
        yScale,
        width,
        height,

        top,
        right,
        bottom,
        left,

        theme,
    }
) => {
    const axes = { top, right, bottom, left }

    positions.forEach(position => {
        const axis = axes[position]

        if (!axis) return null

        const isXAxis = position === 'top' || position === 'bottom'
        const ticksPosition = position === 'top' || position === 'left' ? 'before' : 'after'
        const scale = isXAxis ? xScale : yScale
        const format = getFormatter(axis.format, scale)

        renderAxisToCanvas(ctx, {
            ...axis,
            axis: isXAxis ? 'x' : 'y',
            x: position === 'right' ? width : 0,
            y: position === 'bottom' ? height : 0,
            scale,
            format,
            length: isXAxis ? width : height,
            ticksPosition,
            theme,
        })
    })
}

export const renderGridLinesToCanvas = (ctx, { width, height, scale, axis, values }) => {
    const lines = computeGridLines({ width, height, scale, axis, values })

    lines.forEach(line => {
        ctx.beginPath()
        ctx.moveTo(line.x1, line.y1)
        ctx.lineTo(line.x2, line.y2)
        ctx.stroke()
    })
}
