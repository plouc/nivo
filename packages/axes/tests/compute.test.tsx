import { scaleLinear, scaleOrdinal, scalePoint, scaleBand } from 'd3-scale'
// @ts-ignore
import { computeCartesianTicks } from '../src/compute'

describe('computeCartesianTicks()', () => {
    const ordinalScale = scaleOrdinal([0, 10, 20, 30]).domain(['A', 'B', 'C', 'D'])
    const pointScale = scalePoint().domain(['E', 'F', 'G', 'H']).range([0, 300])
    const bandScale = scaleBand().domain(['I', 'J', 'K', 'L']).rangeRound([0, 400])
    const linearScale = scaleLinear().domain([0, 500]).range([0, 100])

    describe('from linear scale', () => {
        it('should compute ticks for x axis', () => {
            expect(
                computeCartesianTicks({
                    scale: linearScale,
                    axis: 'x',
                })
            ).toMatchSnapshot()
        })

        it('should compute ticks for y axis', () => {
            expect(
                computeCartesianTicks({
                    scale: linearScale,
                    axis: 'y',
                })
            ).toMatchSnapshot()
        })

        it('should allow to customize tick values', () => {
            const tickValues = [10, 20, 30]
            const axis = computeCartesianTicks({
                scale: linearScale,
                tickValues,
                axis: 'x',
            })
            expect(axis.ticks.map(({ value }) => value)).toEqual(tickValues)
        })

        it('should allow to customize tick count', () => {
            const axis = computeCartesianTicks({
                scale: linearScale,
                tickValues: 1,
                axis: 'y',
            })
            expect(axis.ticks.length).toBe(2)
        })
    })

    describe('from ordinal scale', () => {
        it('should compute ticks for x axis', () => {
            expect(
                computeCartesianTicks({
                    scale: ordinalScale,
                    axis: 'x',
                })
            ).toMatchSnapshot()
        })

        it('should compute ticks for y axis', () => {
            expect(
                computeCartesianTicks({
                    scale: ordinalScale,
                    axis: 'y',
                })
            ).toMatchSnapshot()
        })
    })

    describe('from point scale', () => {
        it('should compute ticks for x axis', () => {
            expect(
                computeCartesianTicks({
                    scale: pointScale,
                    axis: 'x',
                })
            ).toMatchSnapshot()
        })

        it('should compute ticks for y axis', () => {
            expect(
                computeCartesianTicks({
                    scale: pointScale,
                    axis: 'y',
                })
            ).toMatchSnapshot()
        })
    })

    describe('from band scale', () => {
        it('ticks should be centered', () => {
            const ticks = computeCartesianTicks({
                scale: bandScale,
                axis: 'x',
            })
            expect(ticks.ticks[0].x).toBe(50)
        })

        it('should compute ticks for x axis', () => {
            expect(
                computeCartesianTicks({
                    scale: bandScale,
                    axis: 'x',
                })
            ).toMatchSnapshot()
        })

        it('should compute ticks for y axis', () => {
            expect(
                computeCartesianTicks({
                    scale: bandScale,
                    axis: 'y',
                })
            ).toMatchSnapshot()
        })
    })

    describe('using canvas engine', () => {
        it('should correctly map textAlign property', () => {
            expect(
                computeCartesianTicks({
                    scale: linearScale,
                    axis: 'y',
                    engine: 'canvas',
                }).textAlign
            ).toBe('right')
            expect(
                computeCartesianTicks({
                    scale: linearScale,
                    axis: 'y',
                    ticksPosition: 'after',
                    engine: 'canvas',
                }).textAlign
            ).toBe('left')
            expect(
                computeCartesianTicks({
                    scale: linearScale,
                    axis: 'x',
                    engine: 'canvas',
                }).textAlign
            ).toBe('center')
        })

        it('should correctly map textBaseline property', () => {
            expect(
                computeCartesianTicks({
                    scale: linearScale,
                    axis: 'x',
                    engine: 'canvas',
                }).textBaseline
            ).toBe('bottom')
            expect(
                computeCartesianTicks({
                    scale: linearScale,
                    axis: 'y',
                    engine: 'canvas',
                }).textBaseline
            ).toBe('middle')
        })
    })
})
