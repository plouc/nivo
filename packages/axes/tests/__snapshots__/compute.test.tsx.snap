// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computeCartesianTicks() from band scale should compute ticks for x axis 1`] = `
{
  "textAlign": "middle",
  "textBaseline": "text-before-edge",
  "ticks": [
    {
      "key": "I",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": "I",
      "x": 50,
      "y": 0,
    },
    {
      "key": "J",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": "J",
      "x": 150,
      "y": 0,
    },
    {
      "key": "K",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": "K",
      "x": 250,
      "y": 0,
    },
    {
      "key": "L",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": "L",
      "x": 350,
      "y": 0,
    },
  ],
}
`;

exports[`computeCartesianTicks() from band scale should compute ticks for y axis 1`] = `
{
  "textAlign": "start",
  "textBaseline": "central",
  "ticks": [
    {
      "key": "I",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": "I",
      "x": 0,
      "y": 50,
    },
    {
      "key": "J",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": "J",
      "x": 0,
      "y": 150,
    },
    {
      "key": "K",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": "K",
      "x": 0,
      "y": 250,
    },
    {
      "key": "L",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": "L",
      "x": 0,
      "y": 350,
    },
  ],
}
`;

exports[`computeCartesianTicks() from linear scale should compute ticks for x axis 1`] = `
{
  "textAlign": "middle",
  "textBaseline": "text-before-edge",
  "ticks": [
    {
      "key": "0",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": 0,
      "x": 0,
      "y": 0,
    },
    {
      "key": "50",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": 50,
      "x": 10,
      "y": 0,
    },
    {
      "key": "100",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": 100,
      "x": 20,
      "y": 0,
    },
    {
      "key": "150",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": 150,
      "x": 30,
      "y": 0,
    },
    {
      "key": "200",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": 200,
      "x": 40,
      "y": 0,
    },
    {
      "key": "250",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": 250,
      "x": 50,
      "y": 0,
    },
    {
      "key": "300",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": 300,
      "x": 60,
      "y": 0,
    },
    {
      "key": "350",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": 350,
      "x": 70,
      "y": 0,
    },
    {
      "key": "400",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": 400,
      "x": 80,
      "y": 0,
    },
    {
      "key": "450",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": 450,
      "x": 90,
      "y": 0,
    },
    {
      "key": "500",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": 500,
      "x": 100,
      "y": 0,
    },
  ],
}
`;

exports[`computeCartesianTicks() from linear scale should compute ticks for y axis 1`] = `
{
  "textAlign": "start",
  "textBaseline": "central",
  "ticks": [
    {
      "key": "0",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": 0,
      "x": 0,
      "y": 0,
    },
    {
      "key": "50",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": 50,
      "x": 0,
      "y": 10,
    },
    {
      "key": "100",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": 100,
      "x": 0,
      "y": 20,
    },
    {
      "key": "150",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": 150,
      "x": 0,
      "y": 30,
    },
    {
      "key": "200",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": 200,
      "x": 0,
      "y": 40,
    },
    {
      "key": "250",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": 250,
      "x": 0,
      "y": 50,
    },
    {
      "key": "300",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": 300,
      "x": 0,
      "y": 60,
    },
    {
      "key": "350",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": 350,
      "x": 0,
      "y": 70,
    },
    {
      "key": "400",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": 400,
      "x": 0,
      "y": 80,
    },
    {
      "key": "450",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": 450,
      "x": 0,
      "y": 90,
    },
    {
      "key": "500",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": 500,
      "x": 0,
      "y": 100,
    },
  ],
}
`;

exports[`computeCartesianTicks() from ordinal scale should compute ticks for x axis 1`] = `
{
  "textAlign": "middle",
  "textBaseline": "text-before-edge",
  "ticks": [
    {
      "key": "A",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": "A",
      "x": 0,
      "y": 0,
    },
    {
      "key": "B",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": "B",
      "x": 10,
      "y": 0,
    },
    {
      "key": "C",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": "C",
      "x": 20,
      "y": 0,
    },
    {
      "key": "D",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": "D",
      "x": 30,
      "y": 0,
    },
  ],
}
`;

exports[`computeCartesianTicks() from ordinal scale should compute ticks for y axis 1`] = `
{
  "textAlign": "start",
  "textBaseline": "central",
  "ticks": [
    {
      "key": "A",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": "A",
      "x": 0,
      "y": 0,
    },
    {
      "key": "B",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": "B",
      "x": 0,
      "y": 10,
    },
    {
      "key": "C",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": "C",
      "x": 0,
      "y": 20,
    },
    {
      "key": "D",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": "D",
      "x": 0,
      "y": 30,
    },
  ],
}
`;

exports[`computeCartesianTicks() from point scale should compute ticks for x axis 1`] = `
{
  "textAlign": "middle",
  "textBaseline": "text-before-edge",
  "ticks": [
    {
      "key": "E",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": "E",
      "x": 0,
      "y": 0,
    },
    {
      "key": "F",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": "F",
      "x": 100,
      "y": 0,
    },
    {
      "key": "G",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": "G",
      "x": 200,
      "y": 0,
    },
    {
      "key": "H",
      "lineX": 0,
      "lineY": 10,
      "textX": 0,
      "textY": 15,
      "value": "H",
      "x": 300,
      "y": 0,
    },
  ],
}
`;

exports[`computeCartesianTicks() from point scale should compute ticks for y axis 1`] = `
{
  "textAlign": "start",
  "textBaseline": "central",
  "ticks": [
    {
      "key": "E",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": "E",
      "x": 0,
      "y": 0,
    },
    {
      "key": "F",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": "F",
      "x": 0,
      "y": 100,
    },
    {
      "key": "G",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": "G",
      "x": 0,
      "y": 200,
    },
    {
      "key": "H",
      "lineX": 10,
      "lineY": 0,
      "textX": 15,
      "textY": 0,
      "value": "H",
      "x": 0,
      "y": 300,
    },
  ],
}
`;
